include/deal.II/multigrid/mg_coarse.h

```diff
@@ -284,6 +333,99 @@ MGCoarseGridApplySmoother<VectorType>::operator()(const unsigned int level,
   coarse_smooth->smooth(level, dst, src);
 }
 
+/* ------------------ Functions for MGCoarseGridApplyPreconditioner ----------*/
+
+template <class VectorType, class PreconditionerType>
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::
+  MGCoarseGridApplyPreconditioner()
+  : preconditioner(0, typeid(*this).name())
+{}
+
+
+
+template <class VectorType, class PreconditionerType>
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::
+  MGCoarseGridApplyPreconditioner(const PreconditionerType &preconditioner)
+  : preconditioner(&preconditioner, typeid(*this).name())
+{}
+
+
+
+template <class VectorType, class PreconditionerType>
+void
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::initialize(
+  const PreconditionerType &preconditioner_)
+{
+  preconditioner = &preconditioner_;
+}
+
+
+
+template <class VectorType, class PreconditionerType>
+void
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::clear()
+{
+  preconditioner = 0;
+}
+
+
+namespace internal
+{
+  namespace MGCoarseGridApplyPreconditioner
+  {
+    template <class VectorType,
+              class PreconditionerType,
+              typename std::enable_if<
+                std::is_same<typename VectorType::value_type, double>::value,
+                VectorType>::type * = nullptr>
```

Comment:
I think it would be nicer if we could introspect whether the `PrecondditionerType` supports `vmult` with `VectorType`. I anticipate possible problems if we add float-based preconditioner classes.